service/storage_proxy.cc

```diff
@@ -5694,27 +5695,19 @@ storage_proxy::query_partition_key_range_concurrent(storage_proxy::clock_type::t
     query::result_merger merger(cmd->get_row_limit(), cmd->partition_limit);
     merger.reserve(exec.size());
 
-    auto f = utils::result_map_reduce(exec.begin(), exec.end(), [timeout] (::shared_ptr<abstract_read_executor>& rex) {
+    auto wrapped_result = co_await utils::result_map_reduce(exec.begin(), exec.end(), [timeout] (::shared_ptr<abstract_read_executor>& rex) {
         return rex->execute(timeout);
     }, std::move(merger));
 
-    return utils::result_futurize_try([&] {
-      return f.then(utils::result_wrap([p,
-            erm, // protects &tm
-            &tm,
-            exec = std::move(exec),
-            results = std::move(results),
-            ranges_to_vnodes = std::move(ranges_to_vnodes),
-            cl,
-            cmd,
-            concurrency_factor,
-            timeout,
-            remaining_row_count,
-            remaining_partition_count,
-            trace_state = std::move(trace_state),
-            preferred_replicas = std::move(preferred_replicas),
-            ranges_per_exec = std::move(ranges_per_exec),
-            permit = std::move(permit)] (foreign_ptr<lw_shared_ptr<query::result>>&& result) mutable {
+    if (!wrapped_result) {
+        auto error = bo::failure(std::move(wrapped_result).assume_error()).error();
+        auto ex = error.as_exception_future().get_exception();
+        p->handle_read_error(std::move(error), true);
+        co_return coroutine::return_exception_ptr(std::move(ex));
```

Comment:
I'm sure this could be improved. Since this function returns a `result<>`, it could be propagated.

But, let's not mix the improvement with this series.

Reply:
I'd do that if it wasn't for the call to `handle_read_error`...

Reply:
Hmm, this isn't so bad:
```c++
            auto error = std::move(wrapped_result).assume_error();
            p->handle_read_error(error.clone(), true);
            co_return error;
```

testing...